src/main/java/com/phamhieu/bookapi/domain/user/UserService.java

```diff
@@ -0,0 +1,96 @@
+package com.phamhieu.bookapi.domain.user;
+
+import com.phamhieu.bookapi.persistence.user.UserStore;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+import javax.xml.bind.DatatypeConverter;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+import static com.phamhieu.bookapi.domain.user.UserError.*;
+
+@Service
+@RequiredArgsConstructor
+public class UserService {
+
+    private final UserStore userStore;
+
+    @Value("${security.encode}")
+    private String appSecurityCode;
+
+    public List<User> findAll() {
+        return userStore.findAll();
+    }
+
+    public User findUserById(final UUID userId) {
+        return userStore.findUserById(userId)
+                .orElseThrow(supplyUserNotFoundById(userId));
+    }
+
+    public List<User> findUserByName(String userName) {
+        final List<User> users = userStore.findByNameContain(userName);
+        if (users.size() == 0) {
+            throw supplyUserNotFoundByName(userName).get();
+        }
+        return userStore.findByNameContain(userName);
+    }
+
+    public User addUser(final User user) throws NoSuchAlgorithmException {
+        checkInformation(user);
+
+        verifyUserIfAvailable(user);
+        final User tempUser = User.builder()
+                .id(user.getId())
+                .username(user.getUsername())
+                .password(hashPassword(user.getPassword()))
+                .firstname(user.getFirstname())
+                .lastname(user.getLastname())
+                .enabled(user.isEnabled())
+                .avatar(user.getAvatar())
+                .roleId(user.getRoleId())
+                .build();
+        return userStore.addUser(tempUser);
+    }
+
+    public User updateUser(final User user, final UUID userId) throws NoSuchAlgorithmException {
+        checkInformation(user);
+        User tempUser = findUserById(userId);
```

Comment:
```suggestion
        final User tempUser = findUserById(userId)
                                                      .withPassword(...)
                                                      .withFirstName(...)
                                                      .withLastName(..)
```

Mình sẽ không cho user đổi username hoặc nếu cho thì phải kiểm tra username mới không trùng với 1 username của 1 user khác